utils/stall_free.hh

```diff
@@ -57,14 +59,176 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept TriviallyClearableSequence =
+    Sequence<T> && std::is_trivially_destructible_v<typename T::value_type> && !HasClearGentlyMethod<typename T::value_type>;
+
+template <typename T>
+concept Container = Iterable<T> && requires (T x, typename T::iterator it) {
+    { x.erase(it) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T>
+concept MapLike = Container<T> && requires (T x) {
+    std::is_same_v<typename T::value_type, std::pair<typename T::key_type, typename T::mapped_type>>;
+};
+
+template <HasClearGentlyMethod T>
+future<> clear_gently(T& o) noexcept;
+
+template <SharedPointer T>
+future<> clear_gently(T& o) noexcept;
+
+template <SmartPointer T>
+requires (!SharedPointer<T>)
+future<> clear_gently(T& o) noexcept;
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <typename T>
+requires (StringLike<T> || TriviallyClearableSequence<T>)
+future<> clear_gently(T& s) noexcept;
+
+template <Sequence T>
+requires (!StringLike<T> && !TriviallyClearableSequence<T>)
+future<> clear_gently(T& v) noexcept;
+
+template <MapLike T>
+future<> clear_gently(T& c) noexcept;
+
+template <Container T>
+requires (!StringLike<T> && !Sequence<T> && !MapLike<T>)
+future<> clear_gently(T& c) noexcept;
+
+namespace internal {
+
+template <typename T>
+concept HasClearGentlyImpl = requires (T x) {
+    { clear_gently(x) } -> std::same_as<future<>>;
+};
+
+template <typename T>
+requires HasClearGentlyImpl<T>
+future<> clear_gently(T& x) noexcept {
+    return utils::clear_gently(x);
+}
+
+// This default implementation of clear_gently
+// is required to "terminate" recursive clear_gently calls
+// at trivial objects
+template <typename T>
+requires (!HasClearGentlyImpl<T>)
+future<> clear_gently(T&) noexcept {
+    return make_ready_future<>();
+}
+
+} // namespace internal
+
+template <HasClearGentlyMethod T>
+future<> clear_gently(T& o) noexcept {
+    return futurize_invoke(std::bind(&T::clear_gently, &o));
+}
+
+template <SharedPointer T>
+future<> clear_gently(T& o) noexcept {
+    if (o.use_count() == 1) {
+        return internal::clear_gently(*o);
+    }
+    return make_ready_future<>();
+}
+
+template <SmartPointer T>
+requires (!SharedPointer<T>)
+future<> clear_gently(T& o) noexcept {
+    return internal::clear_gently(*o);
+}
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept {
+    return do_for_each(a, [] (T& o) {
+        return internal::clear_gently(o);
+    });
+}
+
+// Trivially destructible elements can be safely cleared in bulk
+template <typename T>
+requires (StringLike<T> || TriviallyClearableSequence<T>)
+future<> clear_gently(T& s) noexcept {
+    s.resize(0);
+    return make_ready_future<>();
+}
+
+// Clear the elements gently and destroy them one-by-one
+// in reverse order, to avoid copying.
```

Comment:
I'm curious, why is the "in reverse order" important to specify?

Reply:
Otherwise, if we erase elements from a vector one at a time begin() to end() all remaining elements are copied at each step.

Reply:
but more than the order, the reason this is implemented with `pop_back` and not `erase` is not to require the value_type to be copyable.